dpnp/fft/dpnp_utils_fft.py

```diff
@@ -0,0 +1,212 @@
+#   and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+# THE POSSIBILITY OF SUCH DAMAGE.
+# *****************************************************************************
+
+"""
+Helping functions to implement the FFT interface.
+
+These include assertion functions to validate input array and
+functions with the main implementation part to fulfill the interface.
+The main computational work is performed by enabling FFT functions
+available as a pybind11 extension.
+
+"""
+
+# pylint: disable=protected-access
+# pylint: disable=c-extension-no-member
+# pylint: disable=no-name-in-module
+
+import dpctl
+import dpctl.tensor as dpt
+import dpctl.tensor._tensor_impl as ti
+import numpy
+
+import dpnp
+import dpnp.backend.extensions.fft._fft_impl as fi
+
+from ..dpnp_array import dpnp_array
+from ..dpnp_utils import map_dtype_to_device
+
+__all__ = [
+    "dpnp_fft",
+]
+
+
+def _check_norm(norm):
+    if norm not in (None, "ortho", "forward", "backward"):
+        raise ValueError(
+            f"Invalid norm value {norm} should be None, "
+            '"ortho", "forward", or "backward".'
+        )
+
+
+def _fft(a, norm, is_forward, hev_list, dev_list, axes=None):
+    """Calculates FFT of the input array along the specified axes."""
+
+    index = 0
+    if axes is not None:  # batch_fft
+        len_axes = 1 if isinstance(axes, int) else len(axes)
+        local_axes = numpy.arange(-len_axes, 0)
+        a = dpnp.moveaxis(a, axes, local_axes)
+        a_shape_orig = a.shape
+        local_shape = (-1,) + a.shape[-len_axes:]
+        a = dpnp.reshape(a, local_shape)
+        index = 1
+
+    shape = a.shape[index:]
+    strides = (0,) + a.strides[index:]
+
+    if a.dtype == dpnp.complex64:
+        dsc = fi.Complex64Descriptor(shape)
+    else:
+        dsc = fi.Complex128Descriptor(shape)
+
+    dsc.fwd_strides = strides
+    dsc.bwd_strides = dsc.fwd_strides
+    dsc.transform_in_place = False
+    if axes is not None:  # batch_fft
+        dsc.fwd_distance = a.strides[0]
+        dsc.bwd_distance = dsc.fwd_distance
+        dsc.number_of_transforms = numpy.prod(a.shape[0])
+    dsc.commit(a.sycl_queue)
+
+    res = dpt.usm_ndarray(
```

Comment:
shouldn't we use `dpnp.empty_like` here?

Reply:
Added TODO to replace with `dpnp.empty_like` when its bug is fixed. https://github.com/IntelPython/dpctl/issues/1714

Reply:
It seems OneMKL FFT requires input and output arrays to have the same strides. Default value for `dpnp.empty_like` is `order="C"`. Even if support for `order="K"` is added to `dpnp.empty_like`, it does not guarantee that output has the same strides as input. To make sure output array has the correct strides for FFT operation, it is necessary to use `dpt.usm_ndarray`.